File: src/main/java/org/wysko/midis2jam2/instrument/monophonic/pipe/Ocarina.java
package org.wysko.midis2jam2.instrument.monophonic.pipe;

import com.jme3.math.Quaternion;
import com.jme3.math.Vector3f;
import com.jme3.scene.Node;
import com.jme3.scene.Spatial;
import org.wysko.midis2jam2.Midis2jam2;
import org.wysko.midis2jam2.instrument.LinearOffsetCalculator;
import org.wysko.midis2jam2.instrument.monophonic.HandedClone;
import org.wysko.midis2jam2.midi.MidiChannelSpecificEvent;

import java.util.HashMap;
import java.util.List;

import static org.wysko.midis2jam2.Midis2jam2.rad;

public class Ocarina extends HandedInstrument {
	
	/**
	 * Constructs an ocarina.
	 *
	 * @param context context to midis2jam2
	 */
	public Ocarina(Midis2jam2 context, List<MidiChannelSpecificEvent> events) throws ReflectiveOperationException {
		super(context,
				events,
				OcarinaClone.class,
				new OcarinaHandGenerator()
		);
		
		groupOfPolyphony.setLocalTranslation(45, 47, 20);
		groupOfPolyphony.setLocalRotation(new Quaternion().fromAngles(0, rad(135), 0));
		
	}
	
	static class OcarinaHandGenerator extends HashMap<Integer, HandedClone.Hands> {
		@Override
		public HandedClone.Hands get(Object key) {
			return new HandedClone.Hands(0, ((int) key + 3) % 12);
		}
	}
	
	public class OcarinaClone extends HandedClone {
		public OcarinaClone() {
			super(Ocarina.this, 0);
			highestLevel = new Node();
			Spatial horn = context.loadModel("Ocarina.obj", "Ocarina.bmp", Midis2jam2.MatType.UNSHADED, 0.9f);
			animNode.attachChild(horn);
			highestLevel.attachChild(animNode);
			loadHands();
			for (int i = 0; i < rightHands.length; i++) {
				if (i == 0) rightHands[i].setCullHint(Spatial.CullHint.Dynamic);
				else rightHands[i].setCullHint(Spatial.CullHint.Always);
			}
		}
		
		@Override
		protected void loadHands() {
			rightHands = new Spatial[12];
			for (int i = 0; i < 12; i++) {
				rightHands[i] = context.loadModel("OcarinaHand" + i + ".obj", "hands.bmp", Midis2jam2.MatType.UNSHADED, 0.9f);
			}
			for (Spatial rightHand : rightHands) {
				rightHandNode.attachChild(rightHand);
			}
		}
		
		@Override
		public void tick(double time, float delta) {
			super.tick(time, delta);
			/* Collect note periods to execute */
			if (isPlaying()) {
				animNode.setLocalTranslation(0,
						0, 3 * (float) ((currentNotePeriod.endTime - time) / currentNotePeriod.duration()));
			}
		}
		
		@Override
		protected void moveForPolyphony() {
			offsetNode.setLocalTranslation(10*indexForMoving(),0,0);
		}
	}
}

File: app/src/commonMain/kotlin/org/wysko/midis2jam2/instrument/family/pipe/Ocarina.kt
/*
 * Copyright (C) 2025 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */
package org.wysko.midis2jam2.instrument.family.pipe

import com.jme3.math.Quaternion
import com.jme3.scene.Spatial
import org.wysko.kmidi.midi.event.MidiEvent
import org.wysko.midis2jam2.Midis2jam2
import org.wysko.midis2jam2.instrument.algorithmic.HandPositionFingeringManager
import org.wysko.midis2jam2.instrument.clone.ClonePitchBendConfiguration
import org.wysko.midis2jam2.instrument.clone.CloneWithHands
import org.wysko.midis2jam2.util.Utils.rad
import org.wysko.midis2jam2.world.modelD
import kotlin.time.Duration

/** The Ocarina. */
class Ocarina(context: Midis2jam2, events: List<MidiEvent>) :
    InstrumentWithHands(context, events, OcarinaClone::class, OcarinaHandGenerator()) {

    override val pitchBendConfiguration: ClonePitchBendConfiguration = ClonePitchBendConfiguration(scaleFactor = 0.1f)

    /**
     * The ocarina hand positions are from 0 to 11 and wrap around the octave. So this is easily calculable and doesn't
     * need to be stored in XML.
     */
    internal class OcarinaHandGenerator : HandPositionFingeringManager() {
        override fun fingering(midiNote: Byte): Hands {
            return Hands(0, (midiNote + 3) % 12)
        }
    }

    /** A single ocarina. */
    inner class OcarinaClone : CloneWithHands(this@Ocarina, 0f) {

        override val leftHands: List<Spatial> = listOf()
        override val rightHands: List<Spatial> = List(12) {
            context.modelD("OcarinaHand$it.obj", "hands.bmp")
        }

        override fun tick(time: Duration, delta: Duration) {
            super.tick(time, delta)
            /* Collect note periods to execute */
            if (isPlaying) {
                currentNotePeriod?.let {
                    animNode.setLocalTranslation(
                        0f,
                        0f,
                        3 * ((it.endTime - time) / it.duration).toFloat()
                    )
                }
            }
        }

        override fun adjustForPolyphony(delta: Duration) {
            root.localRotation = Quaternion().fromAngles(0f, rad((17f * indexForMoving()).toDouble()), 0f)
        }

        init {
            loadHands()
            geometry.attachChild(context.modelD("Ocarina.obj", "Ocarina.bmp"))
            highestLevel.setLocalTranslation(0f, 0f, 18f)
        }
    }

    init {
        geometry.setLocalTranslation(32f, 47f, 30f)
        geometry.localRotation = Quaternion().fromAngles(0f, rad(135.0), 0f)
    }
}
